Store Product Image Upload statuses in UserDefaults

pmusolino opened the pull request:
Part 5 of the changes needed for product image upload in the background.

**Before merging this PR, make sure to merge [this one](https://github.com/woocommerce/woocommerce-ios/pull/15257).**


<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

<!-- 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
🛠 Need an AI review?  
Add a comment: `@coderabbitai review`
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
-->

## Description
In this PR:
- I integrated the previously developed `ProductImageStatusStorage` into `ProductImageUploader`, which is the core of the uploader. 
- With the updates in `ProductImageUploader`, the product image statuses will be stored in User Defaults, and stored statuses will become the new source of truth.
- All the changes are under a feature flag `backgroundProductImageUpload`, which currently is enabled only in debug.
- There is also a change in `ProductImage` to have a custom Equatable conformance because I noticed that storing dates while encoding/decoding them could lose some precision, so ignoring fractional seconds makes this comparison accurate.
- This PR updates and implements new unit tests for `ProductImageUploader`, also managing the feature flag.

After this PR, you can expect to see work on real product image uploads in the background.

## Testing information
I tested several scenarios with both the feature flag enabled and disabled. All of them should work as they did previously.

- Uploading one or multiple images to a product and saving them at the end of the upload should correctly save the product with the images included.
- Uploading one or multiple images to a product and saving the product before the upload finishes should result in the images being loaded correctly and attached to the product, including on the web.
- Uploading images to a new product should work as it did before.
- Uploading images to an existing product that already has images should function as it did previously.
- Uploading one image in a product variation should work as it did before.
- Using Network Link Conditioner or deactivating your internet connection during the upload should generate errors, allowing you to retry the image upload.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

pmusolino committed:
Add custom Equatable implementation for ProductImage

pmusolino committed:
feat: use feature flag in `ProductImageUploader` and implement `ProductImageStatusStorage` for storing images statuses in UserDefaults if feature flag is enabled

pmusolino committed:
refactor: Update ProductImageStatusStorage and ProductImageUploader to handle multiple image upload errors

pmusolino committed:
fix: lint warning

pmusolino committed:
Merge branch 'feat/save-product-image-upload-statuses-in-user-defaults-part-4' into feat/save-product-image-upload-statuses-in-user-defaults-part-5

# Conflicts:
#	Networking/Networking/ProductImageInBackground/ProductImageStatusStorage.swift

pmusolino committed:
update: variable name

pmusolino committed:
Merge branch 'feat/save-product-image-upload-statuses-in-user-defaults-part-4' into feat/save-product-image-upload-statuses-in-user-defaults-part-5

pmusolino committed:
feat: new unit tests for `ProductImageUploader` managing `backgroundProductImageUpload` feature flag, and testing that when `backgroundProductImageUpload == true`, `ProductImageStatusStorage` will be used to store image statuses in UserDefaults.

pmusolino commented:
@coderabbitai full review

selanthiraiyan commented on woocommerce-ios:
I think this is an accidental addition and should be removed.

pmusolino replied:
To what are you referring?

selanthiraiyan replied:
Oh, I see a new folder named `woocommerce-ios` added [here](https://github.com/woocommerce/woocommerce-ios/pull/15264/files#diff-3182c6b7fab53dc42ca32e6ac3cc1386d05cc06e7e1441700c7dbde0040b96cb).

pmusolino replied:
Oh, good catch! I think it was generated by mistake while moving all this code in this PR, as it was initially part of a larger branch.

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Media/ProductImageUploaderTests.swift:
```diff
@@ -693,6 +731,154 @@ final class ProductImageUploaderTests: XCTestCase {
         XCTAssertTrue(isNoticeTriggered)
     }
+
+    // MARK: - Tests with Feature Flag Enabled
```
```suggestion
    // MARK: - Tests with background image upload feature flag enabled
```

selanthiraiyan commented on WooCommerce/WooCommerceTests/Mocks/MockProductImageUploader.swift:
```diff
@@ -64,7 +65,12 @@ extension MockProductImageUploader: ProductImageUploaderProtocol {
-        // no-op
+        sendBackgroundUploadNoticeIfNeededWasCalled = true
+
+        // It actually sends a notification if there are active uploads for this key.
+        if activeUploadsKeys.contains(where: { $0 == key }) {
+            noticePresenter.enqueue(notice: Notice(title: "Test Notice"))
+        }
```
Could you kindly explain why do we need to send a notice here? AFAIK, all tests in `ProductImageUploaderTests` pass even without this code.

pmusolino replied:
The test `test_background_upload_notice_is_sent_when_there_are_active_uploads_with_flag_enabled` will not pass without it. This test is not only checking that the method `sendBackgroundUploadNoticeIfNeeded` was called, but also that it correctly sends a notice under the right circumstances (when there are active uploads).

In the real `ProductImageUploader` class, the `sendBackgroundUploadNoticeIfNeeded` method looks at active uploads and only calls `noticePresenter.enqueue(notice:)` if there are uploads for the given key.

The test is verifying this conditional behavior, and checks that:
- When there are NO active uploads → NO notice is sent
- When there ARE active uploads → a notice IS sent

To do that, this test uses `isNoticeTriggered` (set in the `noticePresenter.onNoticeQueued` closure) to know if the notice was actually sent, not just if the method was called. If the mock only tracks that the method was called but doesn't actually enqueue a notice, then `isNoticeTriggered` would remain false even when there are active uploads, causing the test to fail.

selanthiraiyan replied:
Thanks for the response, Paolo!

1. `test_background_upload_notice_is_sent_when_there_are_active_uploads_with_flag_enabled` pass even when I comment out the code sends a notification from `MockProductImageUploader`.
2. I am afraid that we are testing the wrong thing by making `MockProductImageUploader` send a notification. We should instead be testing that `ProductImageUploader` is sending a notification. 

Please correct me if I am misunderstanding this.

pmusolino replied:
That's so strange because, in my case, commenting the code causes it to fail, so I wonder why we are experiencing different behavior.

<img width="706" alt="Screenshot 2025-03-03 at 15 57 16" src="https://github.com/user-attachments/assets/c28dd95b-c025-428f-a0d5-fc7e1367a8f7" />

selanthiraiyan replied:
<img width="1840" alt="image" src="https://github.com/user-attachments/assets/7bded0ae-80d4-4bc8-b346-91bf2d3ed730" />

I commented out only the code that sends out a notification. i.e. I kept the code that updates `sendBackgroundUploadNoticeIfNeededWasCalled` to true. The tests pass because `test_background_upload_notice_is_sent_when_there_are_active_uploads_with_flag_enabled` isn't testing the `noticePresenter`. It only tests the `sendBackgroundUploadNoticeIfNeededWasCalled` property.

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Media/ProductImageUploaderTests.swift:
```diff
@@ -693,6 +731,154 @@ final class ProductImageUploaderTests: XCTestCase {
+        let remoteProductID = productID.id
+        let localID = ProductOrVariationID.product(id: localProductID)
+
+        // When
+        mockImageUploader.replaceLocalID(siteID: siteID,
+                                        localID: localID,
+                                        remoteID: remoteProductID)
+
+        // Then
+        XCTAssertTrue(mockImageUploader.replaceLocalIDWasCalled, "Replace local ID should be called")
```
I don't understand what we are trying to test here. We are firing a method over the `mockImageUploader` and testing that the method is fired. Did you intend to test that storage is updated? 🤔

pmusolino replied:
This is a valid observation; as you can see, storage is not included at all here, so I would say that the name is misleading. I renamed it to `test_replaceLocalID_is_called_when_flag_enabled` 12f4c1b0b96e5824bb7175a666a1a98666e548a7

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Media/ProductImageUploaderTests.swift:
```diff
@@ -693,6 +731,154 @@ final class ProductImageUploaderTests: XCTestCase {
+        mockFeatureFlagService = MockFeatureFlagService(backgroundProductImageUpload: true)
+        let mockImageUploader = MockProductImageUploader()
+
+        // When
+        mockImageUploader.reset()
+
+        // Then
+        XCTAssertTrue(mockImageUploader.resetWasCalled, "Reset should be called")
```
I don't understand why we are firing a method over `MockProductImageUploader` and checking that the method was called. We should instead be creating an instance of `ProductImageUploader` and testing that it clears the storage when `reset` method is called. 

Please correct me if I am missing something.

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Media/ProductImageUploaderTests.swift:
```diff
@@ -693,6 +731,154 @@ final class ProductImageUploaderTests: XCTestCase {
+        XCTAssertEqual(activeUploads.count, 0)
+    }
+
+    func test_background_upload_notice_is_sent_when_there_are_active_uploads_with_flag_enabled() {
+        // Given
+        mockFeatureFlagService = MockFeatureFlagService(backgroundProductImageUpload: true)
+
+        let mockImageUploader = MockProductImageUploader()
+
+        let key = ProductImageUploaderKey(siteID: siteID,
+                                         productOrVariationID: productID,
+                                         isLocalID: false)
+
+        let noticePresenter = MockNoticePresenter()
+
+        // When - No active uploads
+        mockImageUploader.activeUploadsKeys = []
+        mockImageUploader.sendBackgroundUploadNoticeIfNeeded(key: key, using: noticePresenter)
+
+        // Then
+        XCTAssertTrue(mockImageUploader.sendBackgroundUploadNoticeIfNeededWasCalled)
+
+        // Reset flag for next check
+        mockImageUploader.sendBackgroundUploadNoticeIfNeededWasCalled = false
+
+        // When - With active uploads
+        mockImageUploader.activeUploadsKeys = [key]
+        mockImageUploader.sendBackgroundUploadNoticeIfNeeded(key: key, using: noticePresenter)
+
+        // Then
+        XCTAssertTrue(mockImageUploader.sendBackgroundUploadNoticeIfNeededWasCalled)
+    }
```
Same as the other comments about tests. This test should be testing `ProductImageUploader`. Instead, it creates an instance of `MockProductImageUploader` and tests the mock.

selanthiraiyan reviewed (commented):
Thanks for working on this, Paolo!

I left a few questions about the unit tests. Please take a look when you can.

pmusolino committed:
fix: improve error handling for image upload failures in `ProductImageUploader`, considering all the errors and not just the first one.

pmusolino commented:
Thank you for the review, @selanthiraiyan! 🙌 I will respond to your feedback step by step.

I noticed an issue that was addressed in another class and in a previous PR, so I fixed it here as well. I ensured that all error statuses are processed and reported, not just the first one. 3a14b530abddb3243f376cccc7c73ae2fdae07ba

pmusolino committed:
Update WooCommerce/WooCommerceTests/ViewRelated/Products/Media/ProductImageUploaderTests.swift

Co-authored-by: Sharma Elanthiraiyan <sharma.elanthiraiyan@automattic.com>

pmusolino committed:
rename method name in test

itsmeichigo commented on WooCommerce/Classes/ServiceLocator/ProductImageUploader.swift:
```diff
@@ -87,12 +89,52 @@ protocol ProductImageUploaderProtocol {
 /// Supports background image upload and product images update after the user leaves the product form.
 final class ProductImageUploader: ProductImageUploaderProtocol {
+
+    let userDefaultsStatuses: ProductImageStatusStorage?
```
Nit: to make it easier to understand, please consider renaming the property to `imageStatusStorage`. Also, it should be private because it's not meant to be accessible from outside.

itsmeichigo replied:
It's hard to work with this property when it's optional. How about maintaining it without the feature flag checks, and use the feature flag only in getters like for `errors` and `activeUploads` properties?

That way, we can limit the use of the feature flag check and the code would be easier to read/maintain too.

pmusolino replied:
> Nit: to make it easier to understand, please consider renaming the property to imageStatusStorage.

What about just `storage`? We called it `storage` also somewhere else.

pmusolino replied:
> It's hard to work with this property when it's optional. How about maintaining it without the feature flag checks, and use the feature flag only in getters like for errors and activeUploads properties?
> 
> That way, we can limit the use of the feature flag check and the code would be easier to read/maintain too.

Excellent suggestions. I applied the changes here ab508e5dc921e46adc826a92b15ef224f32ca243

itsmeichigo replied:
> What about just `storage`? We called it `storage` also somewhere else.

I think it would not be clear that `storage` is for image statuses.

itsmeichigo reviewed (commented):
I took a quick look at the updates in `ProductImageUploader` and have some comments too:

pmusolino committed:
update: remove optional chaining from `userDefaultsStatuses` in `ProductImageUploader` and always instantiate `ProductImageStatusStorage` in `ProductImageUploader`

pmusolino committed:
- Remove unused code in `MockProductImageUploader.swift`
- Refactor `ProductImageUploaderTests.swift` to use actual `ProductImageUploader` instead of mock
- Update test logic to check actual upload status and actions
- Remove redundant test storage key and update UserDefaults handling in `ProductImageUploaderTests`

pmusolino committed:
revert: submodule pushed by mistake

pmusolino commented:
@selanthiraiyan thanks for all the suggestions! I updated the tests you mentioned since you were right, also removing the unnecessary code in `MockProductImageUploader`, and you can find them in this commit 927e1997c504d196b43ad8d48acc86c891d674f6. I also removed the folder you mentioned (in reality Github doesn't show it correctly, it looks like a folder but in Sourcetree it's a submodule), reverting it here a3f7fd493c3fe940a795e43a5162d0b6762d5bf3, so it's ready for another look.

itsmeichigo commented on WooCommerce/Classes/ServiceLocator/ProductImageUploader.swift:
```diff
@@ -278,16 +352,43 @@ private extension ProductImageUploader {
-                /// When all pending uploads are completed or removed,
-                /// remove the key from active uploads
-                removeProductFromActiveUploads(key: key)
+            if featureFlagService.isFeatureFlagEnabled(.backgroundProductImageUpload) {
+                // Update the states in userDefaultsStatuses
+                self.userDefaultsStatuses.appendStatuses(productImageStatuses, for: key.siteID, productID: key.productOrVariationID)
+            }
+            else {
+                if !activeUploadsPublisher.contains(key), productImageStatuses.hasPendingUpload {
+                    activeUploadsPublisher.append(key)
+                } else if activeUploadsPublisher.contains(key), !productImageStatuses.hasPendingUpload {
+                    /// When all pending uploads are completed or removed,
+                    /// remove the key from active uploads
+                    removeProductFromActiveUploads(key: key)
+                }
```
Since you're maintaining `userDefaultsStatuses` disregarding the feature flag following my suggestion, the feature flag check should be removed here. 

Also, are you sure you want to append every new status in the storage? Consider this case:
- An image is uploaded, you receive the first status being `uploading` for a given product.
- When the upload completes, you get a new status for the same image, this time it's `remote`.

Appending every new status means the storage keeps the `uploading` status even after the upload completes.

So instead, you should replace all the statuses in the storage for a given product with the new ones you received in this observation.

pmusolino replied:
Thank you Huong for the suggestion. 

> Since you're maintaining userDefaultsStatuses disregarding the feature flag following my suggestion, the feature flag check should be removed here.

I kept this logic behind the feature flag because otherwise, it means writing states in User Defaults regardless of the feature flag, which is unnecessary. The change previously suggested was mostly done to avoid unwrapping an optional value.
Wdyt? 

> Also, are you sure you want to append every new status in the storage? Consider this case:
> 
> An image is uploaded, you receive the first status being uploading for a given product.
> When the upload completes, you get a new status for the same image, this time it's remote.
> Appending every new status means the storage keeps the uploading status even after the upload completes.
> 
> So instead, you should replace all the statuses in the storage for a given product with the new ones you received in this observation.

I hope I haven't misunderstood what you want to say, but if you look at the implementation of [appendStatuses](https://github.com/woocommerce/woocommerce-ios/blob/a3f7fd493c3fe940a795e43a5162d0b6762d5bf3/Networking/Networking/ProductImageInBackground/ProductImageStatusStorage.swift#L119) inside `ProductImageStatusStorage`, this method already does what we are trying to do, filtering out the existing statuses for the specific product before adding new ones.

The filters keep everything EXCEPT statuses that match our criteria (keeping statuses from other sites/products). So it's already handling the scenario you described where an uploading status is replaced by a remote status.

If this were not the case, I would also expect that the status of the images uploaded during the application testing would be broken.

itsmeichigo replied:
Thanks for the explanations, sound good to me!

itsmeichigo commented on WooCommerce/Classes/ServiceLocator/ProductImageUploader.swift:
```diff
@@ -302,15 +403,31 @@ private extension ProductImageUploader {
                 if statusUpdatesExcludedProductKeys.contains(key) == false {
-                    errorsSubject.send(infoError)
+                    if featureFlagService.isFeatureFlagEnabled(.backgroundProductImageUpload) {
+                        let failedStatus = ProductImageStatus.uploadFailure(asset: asset,
+                                                                            error: error,
+                                                                            siteID: key.siteID,
+                                                                            productID: key.productOrVariationID)
+                        self.userDefaultsStatuses.updateStatus(failedStatus)
```
Updating the status in the storage is unnecessary here since you were already receiving status updates in `observeStatusUpdates`.

pmusolino replied:
Good catch Huong! This created a duplicated storage update for the same failure event. I tested it and it didn't create race condition or anything, but we definitely update the state unnecessarily (adding more writes to the file system).

So, I updated it here ff2a952b6efcfd26ca952d2573f90e39064b1cba adding also some reminders for the future.

itsmeichigo commented on WooCommerce/Classes/ServiceLocator/ProductImageUploader.swift:
```diff
@@ -302,15 +403,31 @@ private extension ProductImageUploader {
 
     func removeProductFromActiveUploads(key: Key) {
-        activeUploadsPublisher.removeAll(where: { $0 == key })
+        if featureFlagService.isFeatureFlagEnabled(.backgroundProductImageUpload) {
+            userDefaultsStatuses.removeStatus(where: { status in
+                status.siteID == key.siteID &&
+                status.productOrVariationID == key.productOrVariationID &&
+                status.isUploading
+            })
+        } else {
+            activeUploadsPublisher.removeAll(where: { $0 == key })
+        }
```
Same comment as above, since you're maintaining the storage disregarding of the feature flag, you should remove the feature flag check here.

pmusolino replied:
As mentioned above, I kept this logic behind the feature flag because otherwise, it means writing states in User Defaults regardless of the feature flag, which is unnecessary (and it's just an if-else condition). The change previously suggested was mostly done to avoid unwrapping an optional value.
Wdyt?

itsmeichigo replied:
I suppose we'll remove the code on the other side of the feature flag checks, so keeping this for now should be fine 👍

itsmeichigo commented on WooCommerce/Classes/ServiceLocator/ProductImageUploader.swift:
```diff
@@ -248,13 +304,31 @@ final class ProductImageUploader: ProductImageUploaderProtocol {
     private func scheduleUploadInProgressNotificationIfNeeded() {
-        guard !activeUploadsPublisher.isEmpty else { return }
+        if featureFlagService.isFeatureFlagEnabled(.backgroundProductImageUpload) {
+            userDefaultsStatuses.statusesPublisher
+                .map { statuses in
+                    statuses.filter { $0.isUploading }
+                }
+                .sink { uploadingStatuses in
+                    if !uploadingStatuses.isEmpty {
+                        let notification = LocalNotification(scenario: .productImageBackgroundUpload)
+                        let trigger = UNTimeIntervalNotificationTrigger(timeInterval: 1, repeats: false)
+                        Task {
+                            await LocalNotificationScheduler(pushNotesManager: ServiceLocator.pushNotesManager).schedule(notification: notification,
+                                                                                                                         trigger: trigger, remoteFeatureFlag: nil)
+                        }
+                    }
+                }
+                .store(in: &cancellables)
```
I think observing the storage is not necessary here as this method is called in the foreground before the app is sent to the background. 

You should be able to access the statuses directly through the storage's `getAllStatuses`.

pmusolino replied:
Makes sense and good catch. I updated it here: 2607997ee9e1ca6c59d598300d22e50e9dd63c23

itsmeichigo reviewed (changes requested):
Thanks for the updates @pmusolino, I have more comments on the updates in `ProductImageUploader`:

pmusolino committed:
Replace `userDefaultsStatuses` with `imageStatusStorage` throughout `ProductImageUploader`

pmusolino committed:
fix: imageStatusStorage.clearAllStatuses()

pmusolino committed:
- Remove storage update for failed image upload when feature flag is enabled
- Add comments to clarify error handling based on feature flag status

pmusolino committed:
refactor: simplify notification scheduling for background image uploads without the need to observe storage changes.

itsmeichigo commented on WooCommerce/Classes/ServiceLocator/ProductImageUploader.swift:
```diff
@@ -278,16 +350,43 @@ private extension ProductImageUploader {
+                        )
+                        self.errorsSubject.send(errorInfo)
+                    }
+                }
```
I missed this in the last review. Could you please explain why we need to update `errorsSubject` here? Looks like `errorsSubject` is used only in `errors` when the feature flag is disabled, otherwise, `imageStatusStorage.errorsPublisher` is used.

pmusolino replied:
You are right Huong so I completely removed the `observeStatuses()` method.

Then, in the current implementation, the `errorsSubject` is only used when the feature flag is disabled, as you correctly observed.
And in the `observeImageUploads` method:
```
if !self.featureFlagService.isFeatureFlagEnabled(.backgroundProductImageUpload) {
    // Only send the error directly if `backgroundProductImageUpload` feature flag is disabled
    errorsSubject.send(infoError)
}
```
This is correct and intentional. The code I think now, is following a clear pattern:

- When feature flag is disabled: Use `errorsSubject` directly for error reporting
- When feature flag is enabled: Use` imageStatusStorage` for tracking errors, which publishes them through `imageStatusStorage.errorsPublisher`

Changes applied here: 782291eb438a971e4cf2496c1948a385b8936b37

itsmeichigo commented on WooCommerce/WooCommerceTests/Mocks/MockProductImageUploader.swift:
```diff
@@ -15,6 +15,7 @@ final class MockProductImageUploader {
     var startEmittingErrorsWasCalled = false
     var stopEmittingErrorsWasCalled = false
     var resetWasCalled = false
+    var sendBackgroundUploadNoticeIfNeededWasCalled = false
```
This is no longer in use, should we remove it?

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Media/ProductImageUploaderTests.swift:
```diff
@@ -693,6 +729,265 @@ final class ProductImageUploaderTests: XCTestCase {
+
+    // MARK: - Tests with background image upload feature flag enabled
+
+    func test_hasUnsavedChangesOnImages_becomes_false_after_uploading_and_saving_with_flag_enabled() {
+        // Given
+        mockFeatureFlagService = MockFeatureFlagService(backgroundProductImageUpload: true)
+        let mockImageUploader = MockProductImageUploader()
+        let key = ProductImageUploaderKey(siteID: siteID,
+                                         productOrVariationID: productID,
+                                         isLocalID: false)
+
+        // When
+        mockImageUploader.whenHasUnsavedChangesOnImagesIsCalled(thenReturn: true)
+        XCTAssertTrue(mockImageUploader.hasUnsavedChangesOnImages(key: key, originalImages: []))
+
+        // Then
+        mockImageUploader.whenHasUnsavedChangesOnImagesIsCalled(thenReturn: false)
+        XCTAssertFalse(mockImageUploader.hasUnsavedChangesOnImages(key: key, originalImages: [.fake().copy(imageID: 123)]))
+    }
```
I think this test is redundant as you were testing the mock instead of the actual `ProductImageUploader`.

pmusolino replied:
Yes and I agree, this is pretty similar to the one with the feature flag disabled, so I rewrote it using the actual `ProductImageUploader` implementation with the feature flag enabled 0d4be44c1bed58695eed6fa2cf32a7f87dc21831 

Here is what I do:
1. Verifying initial state has no unsaved changes
2. Uploading an image and confirming it creates unsaved changes
3. Saving the product and confirming changes are now saved

Let me know what you think.

itsmeichigo replied:
Thanks! LGTM 👍

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Media/ProductImageUploaderTests.swift:
```diff
@@ -693,6 +729,265 @@ final class ProductImageUploaderTests: XCTestCase {
+        mockFeatureFlagService = MockFeatureFlagService(backgroundProductImageUpload: true)
+
+        _ = CurrentValueSubject<[ProductImageUploaderKey], Never>([])
+        let mockImageUploader = MockProductImageUploader()
+        mockImageUploader.activeUploadsKeys = []
```
This test should be updated to test `ProductImageUploader` instead of the mock.

pmusolino replied:
👍 b952002bafa1d98aa200476f855314a987998729

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Media/ProductImageUploaderTests.swift:
```diff
@@ -693,6 +729,265 @@ final class ProductImageUploaderTests: XCTestCase {
+        }
+    }
+
+    func test_replaceLocalID_is_called_when_flag_enabled() {
```
Could you please explain why this test is necessary? From what I've seen, `replaceLocalID` method is not affected by the feature flag (no change in this PR)

pmusolino replied:
Also if `replaceLocalID` is not affected by the feature flag, when the background upload feature flag is enabled, image status tracking uses the persistent `imageStatusStorage` mechanism rather than just in-memory dictionaries. So I think it's useful to keep this test to verify that:

- The product ID replacement works correctly with the persistent storage mechanism
- The status tracking continues to work properly after the ID change
- The action handler can still access uploading statuses after the ID replacement

So, maybe this is misleading since the test name mention `when_flag_enabled`, but it's testing the integration between `replaceLocalID` and the storage-based implementation that's enabled by the feature flag. The method itself doesn't change behavior, its effects propagate differently through the system when the feature flag is enabled.

itsmeichigo replied:
I read through the implementation of `replaceLocalID` again and still don't see any connection between this method and the status storage. It works solely with `actionHandlersByProduct`, an in-memory dictionary that holds the action handlers by product upload key. Do you think this implementation still works when the app is in the background?

If you see a connection between `replaceLocalID` and the status storage, please show me where the code that indicates that, because I cannot see it.

pmusolino replied:
This will not work when the app is in the background. It will only function when the app is in the foreground to maintain and properly store the current statuses. In the background, I need to use different logic for `replaceLocalID` because `ProductImageUploader` cannot operate while in the background. The idea is to use the same statuses when the app is in the background, fetching them from UserDefaults if there is anything pending. But this test is just to ensure that the status update replacing the local ID will work also with stored statuses (which means that the feature flag is enabled). If you believe it's useless to test that, I can remove it. In any case, at some point, it will remain only a test when the feature flag is removed.

itsmeichigo replied:
Since we have existing tests for `replaceLocalID` already, I think it's better to remove this test when it's unnecessary.

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Media/ProductImageUploaderTests.swift:
```diff
@@ -3,19 +3,39 @@ import Combine
+    override func setUp() {
+        super.setUp()
+        mockFeatureFlagService = MockFeatureFlagService()
+        UserDefaults.standard.removeObject(forKey: "savedProductUploadImageStatuses")
```
IMO it's better to inject the storage into the `ProductImageUploader` and test that the statuses are persisted correctly. We should avoid using `UserDefaults.standard` and use mock instances in unit tests.

pmusolino replied:
Yes, this is definitely a better approach. I applied the changes here (and fixed also a tests that was failing for me locally when running all the tests together): 2a7d9e5614b1ceb24edb0eb3cd3b58605d5e58d6

itsmeichigo reviewed (commented):
Hi Paolo, I have more comments below. Thank you for your patience!

pmusolino committed:
refactor: remove unnecessary status observation for background image uploads

pmusolino committed:
update: removed unused variable

pmusolino committed:
refactor: improve test for image upload handling and unsaved changes verification

pmusolino committed:
refactor: enhance active upload tracking `test_activeUploads_are_tracked_through_storage_with_flag_enabled` using real `ProductImageUploader` instead of a mock.

pmusolino committed:
Refactor: the ProductImageUploader initializer to accept imageStatusStorage as a parameter, and avoid using UserDefaults.standard using mock instances. Additionally, address the failing unit test `test_activeUploads_are_tracked_through_storage_with_flag_enabled`, which has been inconsistent when run alongside other tests.

itsmeichigo commented on WooCommerce/Classes/ServiceLocator/ProductImageUploader.swift:
```diff
@@ -87,12 +89,47 @@ protocol ProductImageUploaderProtocol {
+        if featureFlagService.isFeatureFlagEnabled(.backgroundProductImageUpload) {
+            return imageStatusStorage.errorsPublisher
+                .flatMap { errorItems in
+                    errorItems.publisher
+                        .compactMap { errorItem in
+                            guard let productOrVariationID = errorItem.productOrVariationID,
+                                  let assetType = errorItem.assetType else { return nil }
+                            return ProductImageUploadErrorInfo(
+                                siteID: errorItem.siteID,
+                                productOrVariationID: productOrVariationID,
+                                error: .failedUploadingImage(asset: assetType, error: errorItem.error)
+                            )
+                        }
```
I also missed in the previous reviews that the stream should exclude any keys listed in `statusUpdatesExcludedProductKeys`. That way, we would not display the error notice when an upload fails while the user is on the same product form.

<img src="https://github.com/user-attachments/assets/b57a3aff-7fa9-468e-8aca-4b807dbf8fb3" width=320 />

Notice above that both the error notice and alert are displayed.

pmusolino replied:
Nice catch Huong, I fixed it here 5b719e4e693208eb07bf80d71efa83b2ca51a532 can you please take another look?

itsmeichigo reviewed (commented):
I tested the app with the changes and found that uploading still works as expected. I saw an issue with the error handling though, as described in the comment below.

The PR description didn't mention testing the upload in the background. Are there any more PR coming to support that or can we test background upload in this PR already?

itsmeichigo commented:
I'm also seeing that `test_reset_clears_storage_state_with_flag_enabled() in ProductImageUploaderTests` is failing consistently. Please check the test again, thanks!

pmusolino committed:
Merge branch 'feat/save-product-image-upload-statuses-in-user-defaults-part-4' into feat/save-product-image-upload-statuses-in-user-defaults-part-5

pmusolino committed:
fix: Exclude error updates for products being edited in ProductImageUploader

pmusolino commented:
> The PR description didn't mention testing the upload in the background. Are there any more PR coming to support that or can we test background upload in this PR already?

This is part of following PRs (one is still currently in draft), nothing to test in background here, but thanks for asking!

> I'm also seeing that test_reset_clears_storage_state_with_flag_enabled() in ProductImageUploaderTests is failing consistently. Please check the test again, thanks!

It seems that the test is unreliable: it passes when run alone but fails when executed alongside the other tests. I will investigate to determine what is causing the issue. 👍

pmusolino committed:
fix: Use a unique UserDefaults name for each test in ProductImageUploaderTests

pmusolino committed:
fix: failing unit test because of generic PHAsset

itsmeichigo commented on WooCommerce/Classes/ServiceLocator/ProductImageUploader.swift:
```diff
@@ -87,12 +89,65 @@ protocol ProductImageUploaderProtocol {
+                .eraseToAnyPublisher()
+        } else {
+            return errorsSubject
+                .filter { info in
+                    let key = Key(siteID: info.siteID,
+                                  productOrVariationID: info.productOrVariationID,
+                                  isLocalID: info.productOrVariationID.id == 0)
+                    return !self.statusUpdatesExcludedProductKeys.contains(key)
+                }
```
This filter is redundant because we already check for this in `observeImageUploads`.

pmusolino replied:
I do not think it's redundant because `errorsSubject` can receive error events from other places in `ProductImageUploader` besides `observeImageUploads`. For instance, if you look at the `saveProductImagesWhenNoneIsPendingUploadAnymore` method, it calls `errorsSubject.send(...)` directly when saving fails, without checking `statusUpdatesExcludedProductKeys`. Because of that, the filter on errorSubject acts as a "final gatekeeper" ensuring that errors tied to excluded keys are reliably discarded no matter where in `ProductImageUploader` they're sent.
If the only place errors were published was `observeImageUploads` and it always checked for excluded keys, then the `filter` might be redundant, but given errors can come from more than one place, I believe the filter is necessary to ensure excluded keys never emit errors. Wdyt?

itsmeichigo replied:
I think it was intentional that the save product errors were not filtered out. We were only filtering out upload errors because we also listened to the product form errors. The background saving is handled separately from the product form, so any failures would still be displayed. But this code exists before I joined the project so I'm not entirely sure.

If you feel strongly about hiding the background saving failures when the product form is present, please keep the current change.

itsmeichigo reviewed (commented):
Thanks Paolo for the updates and explanations! I left my last feedback on the threads, please check them out. I'm pre-approving this as the PR should be ready after those last issues are addressed (other than the failing unit test).

pmusolino committed:
Merge branch 'feat/save-product-image-upload-statuses-in-user-defaults-part-4' into feat/save-product-image-upload-statuses-in-user-defaults-part-5

pmusolino committed:
update: removed `test_replaceLocalID_is_called_when_flag_enabled` since it was a redundant test

pmusolino committed:
Merge branch 'trunk' into feat/save-product-image-upload-statuses-in-user-defaults-part-5
